mitodl/ocw-hugo-themes
# replace pdf.js with pdfobject

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-hugo-themes/issues/762
Closes https://github.com/mitodl/ocw-hugo-themes/issues/788

#### What's this PR do?
 - Removes [PDF.js](https://mozilla.github.io/pdf.js/) from the project
 - Adds [PDFObject](https://pdfobject.com/) to the project
 - Moves all code regarding the PDF viewer to the base theme so it can be used anywhere

#### How should this be manually tested?
 - Do a fresh install of the Node dependencies:
   - `rm -rf node_modules`
   - `yarn install`
 - Clone any course from `mitocwcontent` that contains a PDF.  For example, `18.01sc-fall-2010`.  Also, clone `ocw-www`
 - Put the following in your `.env`:
```
COURSE_CONTENT_PATH=/path/to/mitocwcontent/
WWW_CONTENT_PATH=/path/to/mitocwcontent/ocw-www/
OCW_TEST_COURSE=<your test course>
```
 - Spin up the test course with `npm run start:course`
 - Browse to http://localhost:3000/ and find a PDF in your test course (you could also search the `resources` folder in your test course markdown for resources with `resourcetype: Document`)
 - Make sure that the PDF displays properly in every browser you have access to, or doesn't display at all if the browser is not supported.  For example, most mobile browsers will not be supported, and you will only get a download link.
 - Stop your dev server and browse to your `mitocwcontent` folder.  Find a markdown resource file representing a PDF in your course and copy it into the `content/resources` directory in `ocw-www`
 - Spin up `ocw-www` with `npm run start:www`
 - Browse to http://localhost:3000/resources/<your resource file here>
 - Test the PDF in the same manner as above

#### Any background context you want to provide?
Way back in https://github.com/mitodl/hugo-course-publisher/pull/105, we added the PDF.js library to the project with the goal of having a universal way to embed PDFs on pages that worked regardless of the platform.  There were a number of challenges surrounding this.  Mozilla, who maintains PDF.js, offers the open source `pdfjs` Node package on npm, but it does not offer a way to simply load the "example" viewer in your project.  Instead, what they would like people to do is use their library to build their own PDF viewer.

Not being in the business of making PDF viewers, we worked around this by adding a build step to our `package.json` file that would build PDF.js from source and include the example viewer in our project.  Later on, we reduced complexity by removing this build step and simply including a statically built version of the viewer with the course theme.  There were still a number of issues with this:

 - The fully built PDF.js viewer contains a lot of files that need to be copied up to S3 with every theme assets build
 - Cross-domain PDFs could still not be embedded unless the calling domain was allowed via CORS
 - Embedding PDFs in other themes such as `www` would involve moving it to the base theme 
 - PDF.js is a whole other application dependency that needs to be maintained

Enter PDFObject, a simple library consisting of a single Javascript file whos entire job is simply to asses your browser's capabilities and use one of the standard methods for embedding a PDF like `iframe`, `object` or `embed` using a set of rules about browser quirks  that they maintain.  We didn't use PDFObject initially because it was originally desired to have the PDF viewer work even on iOS devices, but those priorities have been reconsidered.  Since PDFObject is a simple JS class with one job, it can easily be used anywhere on the project and requires little to no maintenance.

#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/12089658/177852775-9b29b711-a783-4938-aa07-530da18bdb8c.png)
![image](https://user-images.githubusercontent.com/12089658/177852904-91f92655-5eaf-4f9c-9c11-72074d4af7b4.png)
